Allow exclude unknown property pattern * to match forward slashes

xorye opened the pull request:
Fixes https://github.com/redhat-developer/vscode-quarkus/issues/237 by allowing a pattern terminating with `*` to match everything, including forward slashes

Signed-off-by: David Kwon <dakwon@redhat.com>

xorye commented:
A different solution to this issue that does not change `AntPathMatcher.java` would be to provide:
`org.acme.restclient.CountriesService/mp-rest/url`
`org.acme.restclient.CountriesService/mp-rest/**`
`org.acme.restclient.CountriesService/**`

in the quick fix, instead of providing:
`org.acme.restclient.CountriesService/mp-rest/url`
`org.acme.restclient.*`
`org.acme.*`
`org.*`

angelozerr commented:
The original antpathmatcher is used to match path. In pour case WE want to match a given property which can contrains slash si your solution is fine. But I Wonder if WE could jeep the original antpathmatcher and do some escape before calling it. M'y fear is that we will introduce some bugs By modifing antpathmatcher. In other words is original antpathmatherbcan support match with escaped slash?

xorye commented:
It seems that the original ant path matcher cannot support escaped slash.

The `tokenizePath` method: https://github.com/redhat-developer/quarkus-ls/blob/e085c21c31bbc4711b665f510cd2478281599278/microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/utils/AntPathMatcher.java#L429-L436

will still separate the path based on `/`.

For example, if we call `tokenizePath("a\\/b\\/c")`, the result is `["a\", "b\", "c"]` instead of `["a/b/c"]`

xorye commented:
A different solution that still involves changing ant path matcher that I came up with, is this:

Replace https://github.com/redhat-developer/quarkus-ls/blob/3597f0697e35f36aad8af8cdab023438860e7d8b/microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/utils/AntPathMatcher.java#L231

with:
```
String[] pathDirs = null;
if (Arrays.asList(pattDirs).contains("**")) {
	pathDirs = tokenizePath(path);
} else {
	pathDirs = new String[]{path};
}
```

angelozerr commented:
@xorye the default AntPathMatcher constructor uses '/' as path separator. Are you testing to create the AntPathMatcher with `new AntPathMatcher("")` ?

xorye commented:
In my latest push, I have added more specific tests.

I have also added one more change to `AntPathMatcher.java`, in order for this test case to pass.

The new change:
https://github.com/redhat-developer/quarkus-ls/blob/a423e8f39bb73d87816194396715e5de596257ad/microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/utils/AntPathMatcher.java#L231-L241

The test case for that new change
https://github.com/redhat-developer/quarkus-ls/blob/d369348a1d51393ddf889cf54ebfffdcd2f8c77e/microprofile.ls/com.redhat.microprofile.ls/src/test/java/com/redhat/microprofile/services/ApplicationPropertiesDiagnosticsTest.java#L221-L227

The application properties file for the test case:
```
com.mycompany.remoteServices.MyServiceClient/mp-rest/url/foo=url
com.mycompany.remoteServices.MyServiceClient/mp-rest/uri/bar=uri
com.mycompany.remoteServices.MyOtherClient/mp-rest/url/foo=url
com.mycompany.remoteServices.MyOtherClient/mp-rest/uri/bar=uri
com.mycompany.foo=bar
```
Please let me know if my comments/fix in `AntPathMatcher.java` are not good, I am more than happy to improve the comments/fix.

fbricon commented:
works as advertised. Thanks @xorye!
